Multiple fixes to test_c10d.py.

pritamdamania87 opened the pull request:
Summary:
1) There was a bug in https://github.com/pytorch/pytorch/pull/25012, where the
tests which needed to be skipped for return code checking was incorrect.
2) Added proper setup and teardown for the nccl_error tests.
3) Ensure AssertionError is not ignored for tests that skip return code
checking.

Test Plan: unit tests

Differential Revision: D17003555

mrshenli commented on test/test_c10d.py:
```diff
@@ -2886,30 +2900,25 @@ def _test_nccl_errors_blocking(self, func):
 
     @requires_nccl()
     @skip_if_not_multigpu
-    def test_nccl_errors_blocking_sigstop(self):
```
any reason for removing this test?

pietern replied:
`SIGSTOP` explicitly tests for the timeout condition though, whereas the others test termination.

What about sharing the PIDs through an allgather?

mrshenli commented on test/common_distributed.py:
```diff
@@ -146,8 +150,21 @@ def _join_processes(self, fn):
-        if fn in self.skip_return_code_checks:
+        if fn not in self.skip_return_code_checks:
             self._check_return_codes(elapsed_time)
+        else:
```
shall we swap the order of `if` and `else` so that we don't need to add an additional not?

mrshenli commented on test/common_distributed.py:
```diff
@@ -100,7 +101,10 @@ def wrapper(self):
-                fn(self)
+                try:
+                    fn(self)
+                except AssertionError:
```
what if there is an `RuntimeError`?

mrshenli commented:
CI fails seem relevant:

```
Aug 24 06:09:35 ======================================================================
Aug 24 06:09:35 ERROR: test_broadcast_coalesced_gloo_cpu (__main__.CommTest)
Aug 24 06:09:35 ----------------------------------------------------------------------
Aug 24 06:09:35 Traceback (most recent call last):
Aug 24 06:09:35   File "test_c10d.py", line 2819, in setUp
Aug 24 06:09:35     self.test_nccl_errors_blocking_abort.__wrapped__,
Aug 24 06:09:35 AttributeError: 'function' object has no attribute '__wrapped__'
```

pietern commented on test/test_c10d.py:
```diff
@@ -2793,6 +2793,21 @@ def test_multi_limit_multi_dtype(self):
+        try:
+            func(self)
+        finally:
+            os.environ["NCCL_BLOCKING_WAIT"] = "0"
```
The processes are started fresh for every test. Resetting the env var is not needed.

pietern commented on test/common_distributed.py:
```diff
@@ -146,8 +150,21 @@ def _join_processes(self, fn):
+        """
+        for i, p in enumerate(self.processes):
+            if p.exitcode is None:
+                raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time))
```
This condition is only true if the call to `join` timed out. It won't have terminated here.

pietern replied:
Probably an oversight. Commenting here because the lines changed :)

pietern commented on test/common_distributed.py:
```diff
@@ -146,8 +150,21 @@ def _join_processes(self, fn):
         elapsed_time = time.time() - start_time
-        if fn in self.skip_return_code_checks:
+        if fn not in self.skip_return_code_checks:
             self._check_return_codes(elapsed_time)
```
What about the other exit statuses where you want to skip status checking?

pietern replied:
Nevermind, I thought this was intentional before, but it was an error before and is now fixed.

pietern commented on test/test_c10d.py:
```diff
@@ -2793,6 +2793,21 @@ def test_multi_limit_multi_dtype(self):
+    '''
+        This function is to ensure we run some setup steps before the test
+        method is called on each process.
+    '''
```
AFAIK the convention is to indent the comment block at the same level as the triple quote.

pietern commented:
@pytorchbot rebase this please

pietern commented:
Looks like there is some unrelated failure that made the CUDA tests fail.

pritamdamania committed:
Multiple fixes to test_c10d.py. (#25144)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/25144

1) There was a bug in https://github.com/pytorch/pytorch/pull/25012, where the
tests which needed to be skipped for return code checking was incorrect.
2) Added proper setup and teardown for the nccl_error tests.
3) Ensure AssertionError is not ignored for tests that skip return code
checking.

Test Plan: unit tests

Differential Revision: D17003555

fbshipit-source-id: d1279f87d1a4833596c01fbbcc249fa6a0ada791
